raiden_libs/network/matrix/client.py

```diff
@@ -76,14 +79,15 @@ def _send(self, method, path, *args, **kwargs):
                     raise
                 if time.time() > started + self.retry_timeout:
                     raise
+                last_ex = ex
```

Comment:
shouldn't we chain the exceptions?

Reply:
No need, py3 already keeps the `__context__` in which it happened, so we may just re-raise. Python2 allowed this pattern, of pure `raise` on `else` of `for`, Python3 doesn't keep the exception context outside of the `except` block, but saves it in the `Exception` instance (tested here and seems working)

Reply:
I don't follow.

My understanding is that when `raise` is used inside a `except` block, the exceptions are chained, otherwise they are not. What am I missing?

Reply:
You're missing that, since python3, the `Exception.__traceback__` and `__context__` stores the context of the exception, so no explicit chaining needs to be done, and re-raising an explicit exception instance is needed in a `else` of a `for` block, as it's outside the `except:` block.
Try this:
```
def test():
    raise ValueError('abc')
def test2():
    try:
        test()
    except ValueError as ex:
        return ex  # exception instance with __traceback__ set
def test3():
    ex = test2()
    raise ex
```
You'll notice the last raised `ex` on `test3` keeps the right traceback and context. If you try a simple `raise` in a `else` from `for` block, it won't work, as it's outside a `except:` context, but `raise ex`, if you saved the last exception, will keep track of the context and work as expected.

Reply:
I'm not talking about `when a transaction can be raise`, I'm talking about `chaining transactions`.

Since the snippet does multiple retries in a loop, and the code only raises the last exception, we are missing context information from the previous loops.

Reply:
This is intended, I only want to raise the last exception, **iff** the `retry_delay` iterable finished before `retry_timeout`, like the exception happened in the first call, as this method wraps an HTTP request (and the http request error is raised in retry/timeout case)